### DISCUSSION TITLE
IDEA indexing

### DISCUSSION THREAD

**Discussion by Adam-Vandervorst** [raquo/Laminar#112] [Q&A]
IDEA does not give me code insights into Laminar.

My setup:
IDEA: 2021.2.3 
Scala: 3.1.2-RC1-bin-20211124-1f48de1-NIGHTLY
SBT: 1.5.5
Dependencies: just `"com.raquo" %%% "laminar" % "0.13.1"`

Details:
`import com.raquo.laminar.api.L.{*, given}`
does work with fastLinkJS, but gives no insights beyond the `api` level.
`import com.raquo.laminar.nodes.RootNode`
does give me insights from a `.tasty` file.

I'd say sources of some sort are missing, or something is not loading.
Would love to try out Laminar and insights helps exploring the functions and signatures.

**Comment by yurique** [comment]
This is with scala3, right?
Last time I tried it, barely anything worked in IDEA with scala3 :'(

**Reply by raquo** [reply]
I have to agree. Scala 3 support is so broken / incomplete in IntelliJ, I'm still stuck with Scala 2.

For Laminar I'm using a very standard sbt-release config, I doubt that the problem is on Laminar's end. You can try asking in [their gitter](https://gitter.im/JetBrains/intellij-scala) and/or filing it as a bug in jetbrains youtrack.

**Reply by Adam-Vandervorst** [reply]
Ha I have been using Scala 3 in IDEA for the past 7 months and it's running decently well now.
I don't think it's on Laminar's end, but maybe you had found a way around the particular issue occuring here.

Will definitely reach out, thanks!
Regarding my current Scala 3 project, am I correct in concluding it's not the right time to use Laminar?

**Comment by raquo** [comment]
Why not? I personally am waiting for IntellijJ to sort out their issues,
but Laminar has supported Scala 3 natively for months now.

**Reply by Adam-Vandervorst** [reply]
That's why I'm asking -- good to hear, the project looks like a perfect fit.
I just tested `import com.raquo.laminar.api.Laminar.{*, given}` and that does given me code insights :tada:.
That leads me to believe the inspection doesn't carry through all (any?) information from `val L: com.raquo.laminar.api.Laminar.type = ???`.

**Reply by Adam-Vandervorst** [reply]
Hmm the code insights work but the compiler rejects it with
```scala
[error] 1 |import com.raquo.laminar.api.Laminar.{*, given}
[error]   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]   |object Laminar cannot be accessed as a member of com.raquo.laminar.api.type from module class <empty>.
```

**Reply by Adam-Vandervorst** [reply]
> code insights work

This is an overstatement, as it does highlighting and typing correctly, but it can't see the members.
Concretely
```scala
controlled(value <-- inp, onInput.mapToValue.filter(stateMachine.currentlyValid) --> inp)
```
recognizes the arguments for `controlled` and the signature for `value` and `onInput`, but does not see `.mapToValue`.
Maybe it doesn't see what's behind the `.tasty` definitions.

**Comment by raquo** [comment]
Any IntelliJ issues you run into with Scala 3 will surface easily with
other Scala.js libraries too, or if not the same exact issue, then
something else will pop up. At least that's been my experience.

**Comment by raquo** [comment]
Well, that's to be expected, you should import all that stuff either from
`api` or from `api.L` as mentioned in the docs, not from `api.Laminar` –
that object is private on purpose.

You should look at errors given to you by the real Scala compiler, not by
IntelliJ. Normally that wouldn't be an issue but since their Scala 3
support is not complete you have to account for that.

**Reply by Adam-Vandervorst** [reply]
That makes sense, just reporting it does have some insights into the types that way, over the `api` and `api.L`.

**Comment by raquo** [comment]
Btw try the new IntelliJ 2021.3 version, among the fixes in their changelog:

> In addition, the TASTy reader can now parse package objects

Seems relevant.

--Nikita

**Reply by Adam-Vandervorst** [reply]
That improved something: `L.{*, given}` now provides the same insights as `Laminar.{*, given}`.
So still nothing for `.mapToValue` (or anything but the top level definitions), but that's a start!
Do you see a workaround for the `.tasty` files? I'm considering making it a submodule as this exposes the `.scala` files and building Laminar looks relatively straightforward.

**Reply by raquo** [reply]
Sadly I don't know much about Scala build tools and internals like tasty files
